### DISCUSSION TITLE
command not found: testground

### DISCUSSION THREAD

**Discussion by ebma** [testground/testground#1359] [Q&A] [answered]
Hey there. I followed the steps in the getting started page but after cloning and running `make install` in the testground repo, I can't run `testground daemon` because the `testground` command is not available. Am I missing something here?

 I checked and saw that the `make` command is building some images and I also used docker to run the `iptestground/testground` image and saw that the `testground` command is available there. But that's not how I am supposed to use it or is it? I'm a bit lost and would appreciate any help.

**Comment by ebma** [comment]
I build the testground executable with the following command that is contained in the `Dockerfile.testground`:
```
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X github.com/testground/testground/pkg/version.GitCommit=${TG_VERSION}" -o testground
```

However, I'm still not sure whether this is the right approach or if this is a bug. In case it matters, I'm running manjaro linux with kernel `5.15.48-1-MANJARO x86_64`.

**Comment by mxinden** [comment]
> I can't run `testground daemon` because the `testground` command is not available. Am I missing something here?

Is your Golang `bin` directory (e.g. `~/go/bin`) parth of your path (`echo $PATH`)?

**Reply by ebma** [reply]
It was not part of `$PATH` so I added it but it didn't change the result of `make install`.
In case it is relevant, here are my logs when I run `make install`:
```
(base) ➜  testground git:(master) ✗ make install
go install -ldflags "-X github.com/testground/testground/pkg/version.GitCommit=`git rev-list -1 HEAD`" .
docker build --build-arg TG_VERSION=`git rev-list -1 HEAD` -t iptestground/testground:edge -f Dockerfile.testground .
Sending build context to Docker daemon  67.68MB
Step 1/16 : ARG GO_VERSION=1.16
Step 2/16 : FROM golang:${GO_VERSION}-buster
 ---> 27fe614eb0bc
Step 3/16 : COPY /go.mod /go.mod
 ---> Using cache
 ---> 383226e5d647
Step 4/16 : RUN cd / && go mod download
 ---> Using cache
 ---> 456818dfc735
Step 5/16 : COPY . /
 ---> Using cache
 ---> 2dca3b38e074
Step 6/16 : ARG TG_VERSION
 ---> Using cache
 ---> 22f24dfd7ea8
Step 7/16 : RUN cd / && CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X github.com/testground/testground/pkg/version.GitCommit=${TG_VERSION}" -o testground
 ---> Using cache
 ---> b4531532f0b8
Step 8/16 : FROM golang:${GO_VERSION}-buster
 ---> 27fe614eb0bc
Step 9/16 : RUN mkdir -p /usr/local/bin
 ---> Using cache
 ---> c862e9a71851
Step 10/16 : COPY --from=0 /testground /testground
 ---> Using cache
 ---> c7497df61e81
Step 11/16 : COPY --from=0 /static /static
 ---> Using cache
 ---> 5761982c1a11
Step 12/16 : COPY --from=0 /tmpl /tmpl
 ---> Using cache
 ---> 0ed43bdb619e
Step 13/16 : ENV PATH="/:/usr/local/bin:${PATH}"
 ---> Using cache
 ---> e92a8610c8dc
Step 14/16 : EXPOSE 8042
 ---> Using cache
 ---> 5423970d0b0b
Step 15/16 : WORKDIR "/"
 ---> Using cache
 ---> 15baa27cc29b
Step 16/16 : ENTRYPOINT [ "/testground", "--vv", "daemon" ]
 ---> Using cache
 ---> 6cb60dac621a
Successfully built 6cb60dac621a
Successfully tagged iptestground/testground:edge
docker build --build-arg TG_VERSION=`git rev-list -1 HEAD` -t iptestground/sidecar:edge -f Dockerfile.sidecar .
Sending build context to Docker daemon  67.68MB
Step 1/14 : ARG GO_VERSION=1.16
Step 2/14 : FROM golang:${GO_VERSION}-buster
 ---> 27fe614eb0bc
Step 3/14 : COPY /go.mod /go.mod
 ---> Using cache
 ---> 383226e5d647
Step 4/14 : RUN cd / && go mod download
 ---> Using cache
 ---> 456818dfc735
Step 5/14 : COPY . /
 ---> Using cache
 ---> 2dca3b38e074
Step 6/14 : ARG TG_VERSION
 ---> Using cache
 ---> 22f24dfd7ea8
Step 7/14 : RUN cd / && CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X github.com/testground/testground/pkg/version.GitCommit=${TG_VERSION}" -o testground
 ---> Using cache
 ---> b4531532f0b8
Step 8/14 : FROM debian:buster
 ---> 354ff99d6bff
Step 9/14 : RUN apt update && apt install -y iptables
 ---> Using cache
 ---> fea77d1657fa
Step 10/14 : RUN mkdir -p /usr/local/bin
 ---> Using cache
 ---> 5b91bce0a8bc
Step 11/14 : COPY --from=0 /testground /usr/local/bin/testground
 ---> Using cache
 ---> 94f34e89cab5
Step 12/14 : ENV PATH="/usr/local/bin:${PATH}"
 ---> Using cache
 ---> 6ef5951cdec1
Step 13/14 : EXPOSE 6060
 ---> Using cache
 ---> 3e6a3a2658ed
Step 14/14 : ENTRYPOINT [ "/usr/local/bin/testground" ]
 ---> Using cache
 ---> 858fdd667523
Successfully built 858fdd667523
Successfully tagged iptestground/sidecar:edge
docker pull iptestground/sync-service:latest
latest: Pulling from iptestground/sync-service
Digest: sha256:3fcb7613e78ecfcd83ebc11d83eb449865fe7fca2d8423e3818f50fbf821c744
Status: Image is up to date for iptestground/sync-service:latest
docker.io/iptestground/sync-service:latest
```

Everything seems to be successful. Maybe I am just misunderstanding the purpose of it. 

To clarify: I would expect that if I run `make install`, I get an executable called `testground` in the root of the `testground` repository. But as far as I can tell, `make install` just creates some docker images and one of them includes the `testground` executable. Is this correct and intended? I'm just curious because the ["Getting Started"](https://docs.testground.ai/getting-started) things in the docs suggest otherwise.. I'm not super familiar with Dockerfiles, to be honest, so maybe I'm just misunderstanding something.

**Comment by laurentsenta** [comment, marked as answer]
Hi @ebma, thanks for reaching out. @mxinden is right. That looks like an issue with your PATH.

The first step in `make install` is a `go install` (first line in your logs).

That command builds the testground binary and installs it. Installation and tweaking your PATH is not testground-related but part of the [go build tools](https://go.dev/ref/mod#go-install). We don't distribute binaries yet, so you're in "go developer land" here.

I recommend you double-check your setup to make sure your go installation is correct:

- https://go.dev/doc/install (step 2 might be the one you're missing)
- https://go.dev/ref/mod#go-install
- https://go.dev/doc/tutorial/compile-install

If you want to build in the testground directory, try:

```sh
export GOBIN="${PWD}"
make install # or make goinstall to rebuild only the binary
```

(remember, the binary won't be in your PATH, you'll have to invoke it with `./testground`).

**Reply by ebma** [reply]
Thanks a lot. I installed go with pacman originally but re-installed with yours docs. I didn't know that the `go install` command would put the binary in a specific directory. But it works now, so thanks for your help. :+1:
